src/MoBi.Core/Services/QuantityValueInSimulationChangeTracker.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using MoBi.Core.Domain.Model;
+using MoBi.Core.Events;
+using OSPSuite.Core.Domain;
+using OSPSuite.Core.Domain.Builder;
+using OSPSuite.Utility.Events;
+
+namespace MoBi.Core.Services
+{
+   public interface IQuantityValueInSimulationChangeTracker
+   {
+      void TrackChanges(IQuantity quantity, IMoBiSimulation simulation, Action changeAction);
+   }
+
+   public class QuantityValueInSimulationChangeTracker : IQuantityValueInSimulationChangeTracker
+   {
+      private readonly IQuantityToParameterValueMapper _quantityToParameterValueMapper;
+      private readonly IEventPublisher _eventPublisher;
+
+      public QuantityValueInSimulationChangeTracker(IQuantityToParameterValueMapper quantityToParameterValueMapper, IEventPublisher eventPublisher)
+      {
+         _quantityToParameterValueMapper = quantityToParameterValueMapper;
+         _eventPublisher = eventPublisher;
+      }
+
+      public void TrackChanges(IQuantity quantity, IMoBiSimulation simulation, Action changeAction)
+      {
+         storeOriginalQuantityValue(quantity, simulation);
+
+         changeAction();
+
+         checkForOriginalValueRestored(quantity, simulation);
+      }
+
+      private void storeOriginalQuantityValue(IQuantity quantity, IMoBiSimulation simulation)
+      {
+         simulation.AddOriginalQuantityValue(_quantityToParameterValueMapper.MapFrom(quantity));
```

Comment:
shoy;d thid method take a PV instead of the quantity?
That way you don't have to create a new PV twice every single time (in the store and in the check)